share/availability/discovery/discovery.go

```diff
@@ -96,26 +104,35 @@ func (d *Discovery) WithOnPeersUpdate(f OnUpdatedPeers) {
 
 // handlePeersFound receives peers and tries to establish a connection with them.
 // Peer will be added to PeerCache if connection succeeds.
-func (d *Discovery) handlePeerFound(ctx context.Context, topic string, peer peer.AddrInfo) {
+func (d *Discovery) handlePeerFound(ctx context.Context, peer peer.AddrInfo, cancelOngoing context.CancelFunc) {
 	if peer.ID == d.host.ID() || len(peer.Addrs) == 0 || d.set.Contains(peer.ID) {
 		return
 	}
-	err := d.set.TryAdd(peer.ID)
+
+	ctx, cancel := context.WithTimeout(ctx, d.dialTimeout)
+	defer cancel()
+	err := d.connector.Connect(ctx, peer)
+	// we don't want to add backoff when the context is canceled.
 	if err != nil {
-		log.Debug(err)
+		if errors.Is(err, context.Canceled) || errors.Is(err, routing.ErrNotFound) {
+			d.connector.RemoveBackoff(peer.ID)
+		}
 		return
 	}
 
-	err = d.connector.Connect(ctx, peer)
+	if uint(d.set.Size()) >= d.peersLimit {
+		log.Debugw("peer limit reached", "count", d.set.Size(), "peer", peer.ID)
+		cancelOngoing()
+	}
+	err = d.set.TryAdd(peer.ID)
```

Comment:
I think previously add was before connect to mitigate potential shadow write issue. Example of such issue could this example order of operations:
1. (line114) Peer connected
2. (line182) disconnect event happen
3. (line183) Peer is not in the limited set, so noop. 
4. (line127) Peer added to the pool

If this order happens, peer will land in limited set even it is already disconnected. With previous approach (add->try connect-> remove if fail) this would not be a problem.

There is clearly a benefit of doing add after connect to, but could you please resolve shadow write?

Reply:
That is an interesting point and we will think about it. Maybe you have some ideas as well

Reply:
Well, I think the only correct solution is to keep logic that adds and removes peers synchronized and, ideally, in the same thread. There always will be such races between Add and Remove depending on where event happens.

Reply:
This example is actually one of the reasons behind Rob Pike's quote "Don't communicate by sharing memory; share memory by communicating

Reply:
i didn't think long, so my suggestion could be not optimal. One solution to this corner case that comes to my mind is to track peers that are currently trying to connect in another set.

Reply:
I uploaded the solution